compiler/runtime/OMRCodeCacheManager.cpp

```diff
@@ -315,13 +315,53 @@ OMR::CodeCacheManager::unreserveCodeCache(TR::CodeCache *codeCache)
 // A compThreadID of -1 means unknown. This ID will be written into the code cache
 // The ID of the thread that last reserved the cache will remain written after the
 // reservation is over. This will allow us to implement some affinity.
+//
+// If TR_RetryCodeCacheAllocAndIgnoreKind is set, this method will retry the
+// reservation ignoring the kind; however, on the retry it will skip trying to
+// allocate a new code cache. This is because, on the first attempt, if we
+// failed to reserve a new code cache, we would try to allocate a new code
+// cache. If we're retrying then that means that both the reservation and
+// allocation failed. As such, there is only value in trying to find an
+// existing code cache to reserve.
 TR::CodeCache *
 OMR::CodeCacheManager::reserveCodeCache(bool compilationCodeAllocationsMustBeContiguous,
                                       size_t sizeEstimate,
                                       int32_t compThreadID,
                                       int32_t *numReserved,
                                       TR::CodeCacheKind kind)
    {
+   TR::CodeCache *codeCache = NULL;
+
+   bool retry = false;
+   do
+      {
+      codeCache = reserveCodeCacheImpl(compilationCodeAllocationsMustBeContiguous,
+                                       sizeEstimate, compThreadID, numReserved,
+                                       kind, retry);
+
+      retry =
+         // option to retry is enabled
+         TR::Options::getCmdLineOptions()->getOption(TR_RetryCodeCacheAllocAndIgnoreKind)
+         // only retry once
+         && (retry == false)
+         // failed to reserve or allocate a code cache
+         && !codeCache
+         // the code cache is not full
+         && !self()->codeCacheFull();
+      }
+   while(retry);
```

Comment:
As this loop only runs one or two iterations, I wonder whether it would be easier to read by getting rid of the loop entirely:

```
   codeCache = reserveCodeCacheImpl(compilationCodeAllocationsMustBeContiguous,
                                          sizeEstimate, compThreadID, numReserved,
                                          kind, false /* ignoreKindAndSkipAllocate */);
   if (!codeCache
         && !self()->codeCacheFull()
         && TR::Options::getCmdLineOptions()->getOption(TR_RetryCodeCacheAllocAndIgnoreKind())
      codeCache = reserveCodeCacheImpl(compilationCodeAllocationsMustBeContiguous,
                                       sizeEstimate, compThreadID, numReserved,
                                       kind, true /* ignoreKindAndSkipAllocate */);
```

Reply:
If you keep the do-while loop, may I suggest changing `(retry == false)` to `!retry` and adding a space between the keyword and left parenthesis in `while(retry)`?

Reply:
Yeah I can get rid of the loop; I added it because I wasn't sure if deduplicating code was preferred or not; I don't have any strong feelings either way.